Ввести построчно четвёрки вида «число число число слово», где первые три числа — это координаты галактики по имени «слово» (некоторые галактики могут называться одинаково, но координаты у всех разные). Последняя строка ввода не содержит пробелов и не учитывается. Вывести в алфавитном порядке имена любых двух наиболее удалённых друг от друга галактик.

In [144]:
x = str(input())
d = {}
while ' ' in x:
    x = x.split()
    d[(float(x[0]), float(x[1]), float(x[2]))] = x[3]
    x = str(input())

35.764 -797.636 -770.320 almost
88.213 -61.688 778.457 gene
-322.270 -248.555 -812.730 trend
721.262 630.355 968.287 dow
-895.519 -970.173 97.282 non
-561.036 -350.840 -723.149 disco
-151.546 -900.962 -658.862 bidder
-716.197 478.576 -695.843 hawaii
-744.664 -173.034 -11.211 sad
-999.968 990.467 650.551 erik
.


In [190]:
import time
s = time.time()
max_d = 0
for i in range(len(d)):
    a = list(d.keys())[i]
    for j in range(i, len(d)):
        b = list(d.keys())[j]
# for a in list(d.keys()):
#     for b in list(d.keys()):
        dist = (a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2
        if dist > max_d:
            max_d = dist
            max_couple = [d[a], d[b]]
max_couple = sorted(max_couple)
print(max_couple[0], max_couple[1])
e = time.time()
print((e-s) * 1000)

almost erik
0.9539127349853516


In [150]:
d

{(35.764, -797.636, -770.32): 'almost',
 (88.213, -61.688, 778.457): 'gene',
 (-322.27, -248.555, -812.73): 'trend',
 (721.262, 630.355, 968.287): 'dow',
 (-895.519, -970.173, 97.282): 'non',
 (-561.036, -350.84, -723.149): 'disco',
 (-151.546, -900.962, -658.862): 'bidder',
 (-716.197, 478.576, -695.843): 'hawaii',
 (-744.664, -173.034, -11.211): 'sad',
 (-999.968, 990.467, 650.551): 'erik'}

almost erik

35.764 -797.636 -770.320 almost
88.213 -61.688 778.457 gene
-322.270 -248.555 -812.730 trend
721.262 630.355 968.287 dow
-895.519 -970.173 97.282 non
-561.036 -350.840 -723.149 disco
-151.546 -900.962 -658.862 bidder
-716.197 478.576 -695.843 hawaii
-744.664 -173.034 -11.211 sad
-999.968 990.467 650.551 erik
.